docs/core/diagnostics/diagnosticsource-diagnosticlistener.md

```diff
@@ -0,0 +1,343 @@
+        {
+            _onNext = onNext ?? new Action<T>(_ => { });
+            _onCompleted = onCompleted ?? new Action(() => { });
+        }
+
+        public void OnCompleted() { _onCompleted(); }
+        public void OnError(Exception error) { }
+        public void OnNext(T value) { _onNext(value); }
+
+        private Action<T> _onNext;
+        private Action _onCompleted;
+    }
+    
+    Action<DiagnosticListener> onNewListener = delegate (DiagnosticListener listener)
+    {
+        Console.WriteLine($"New Listener discovered: {listener.Name}");
+        if (listener.Name == "System.Net.Http")
+        {
+            // Here is where we put code to subscribe to the Listener.
+        }
+    };
+    IObserver<DiagnosticListener> observer = new Observer<DiagnosticListener>(onNewListener, null);
+    //when a listener is created, invoke the Observer onNext function which calls the delegate
+    listenerSubscription = DiagnosticListener.AllListeners.Subscribe(observer);
+
+    // Typically you leave the listenerSubscription subscription active forever.
+    // However when you no longer want your callback to be called, you can
+    // call listenerSubscription.Dispose() to cancel your subscription to the IObservable.
+```
+
+This code creates a callback delegate and using the `AllListeners.Subscribe` method requests
+that the delegate be called for every active `DiagnosticListener` in the system. The decision of whether or not to subscribe to the listener
+is made by inspecting its name. The code above is looking for our 'System.Net.Http' listener that we created previously.
+
+Like all calls to `Subscribe()`, this one returns an `IDisposable` that represents the subscription itself.
+Callbacks will continue to happen as long as nothing calls `Dispose()` on this subscription object.
+The above code never calls `Dispose()`, so it will receive callbacks forever.
+
+When you subscribe to `AllListeners`, you get a callback for ALL ACTIVE `DiagnosticListeners`.
+Thus, upon subscribing, you get a flurry of callbacks for all existing `DiagnosticListeners`, and as new ones
+are created, you receive a callback for those as well. You receive a complete list of everything it is possible
+to subscribe to.
+
+#### Subscribing to DiagnosticListeners
+
+A `DiagnosticListener` implements the `IObservable<KeyValuePair<string, object>>` interface, so you can
+call `Subscribe()` on it as well. We show how to fill out the previous example:
+
+```C#
+    static IDisposable networkSubscription = null;
+    static IDisposable listenerSubscription;
+    Action<KeyValuePair<string, object>> onMessage = delegate (KeyValuePair<string, object> message)
+    {
+        Console.WriteLine($"Message received: {message.Key}: {message.Value}");
+    };
+    Action<DiagnosticListener> onNewListener = delegate (DiagnosticListener listener)
+    {
+        if (listener.Name == "System.Net.Http")
+        {
+          lock(allListeners)
```

Comment:
Any locking in prescribed code snippets should be explained.

Reply:
@josalem There is something in the text that describes the locking. Do you think something more than that is needed?

Reply:
I would refactor this snippet to include the _entire_ class definition, so it is obvious where `allListeners` comes from. Right now, it is a variable that isn't defined in the scope of the snippet.